shared/src/balancer/logexpmath.rs

```diff
@@ -0,0 +1,522 @@
+        return Err(Errors::YOutOfBounds);
+    };
+
+    let mut logx_times_y = if *LN_36_LOWER_BOUND < x_int256 && x_int256 < *LN_36_UPPER_BOUND {
+        let ln_36_x = _ln_36(x_int256);
+        (ln_36_x / *ONE_18) * y_int256 + ((ln_36_x % *ONE_18) * y_int256) / *ONE_18
+    } else {
+        _ln(x_int256) * y_int256
+    };
+    logx_times_y /= *ONE_18;
+
+    if !(*MIN_NATURAL_EXPONENT <= logx_times_y && logx_times_y <= *MAX_NATURAL_EXPONENT) {
+        return Err(Errors::ProductOutOfBounds);
+    }
+
+    exp(logx_times_y).map(|v| v.into_raw())
+}
+
+fn exp(mut x: I256) -> Result<I256, Errors> {
+    if !(x >= *MIN_NATURAL_EXPONENT && x <= *MAX_NATURAL_EXPONENT) {
+        return Err(Errors::InvalidExponent);
+    }
+
+    if x < I256::zero() {
+        return Ok((*ONE_18 * *ONE_18) / exp(-x)?);
+    }
+
+    let first_an;
+    if x >= constant_x_18(0) {
+        x -= constant_x_18(0);
+        first_an = constant_a_18(0);
+    } else if x >= constant_x_18(1) {
+        x -= constant_x_18(1);
+        first_an = constant_a_18(1);
+    } else {
+        first_an = 1.into();
+    }
+
+    x *= 100.into();
+
+    let mut product = *ONE_20;
+    for i in 2..=9 {
+        if x >= constant_x_20(i) {
+            x -= constant_x_20(i);
+            product = (product * constant_a_20(i)) / *ONE_20;
+        }
+    }
+
+    let mut series_sum = *ONE_20;
+    let mut term;
+
+    term = x;
+    series_sum += term;
+
+    for i in 2..=12 {
+        term = ((term * x) / *ONE_20) / i.into();
+        series_sum += term;
+    }
+
+    Ok((((product * series_sum) / *ONE_20) * first_an) / 100.into())
+}
+
+fn _ln(mut a: I256) -> I256 {
+    if a < *ONE_18 {
+        return -_ln((*ONE_18 * *ONE_18) / a);
+    }
+
+    let mut sum = I256::zero();
+    for i in 0..=1 {
+        if a >= constant_a_18(i) * *ONE_18 {
+            a /= constant_a_18(i);
+            sum += constant_x_18(i);
+        }
+    }
+
+    sum *= 100.into();
+    a *= 100.into();
+
+    for i in 2..=11 {
+        if a >= constant_a_20(i) {
+            a = (a * *ONE_20) / constant_a_20(i);
+            sum += constant_x_20(i);
+        }
+    }
+
+    let z = ((a - *ONE_20) * *ONE_20) / (a + *ONE_20);
+    let z_squared = (z * z) / *ONE_20;
+
+    let mut num = z;
+    let mut series_sum = num;
+
+    for i in (3..=11).step_by(2) {
+        num = (num * z_squared) / *ONE_20;
+        series_sum += num / i.into();
+    }
+
+    series_sum *= 2.into();
+
+    (sum + series_sum) / 100.into()
+}
+
+fn _ln_36(mut x: I256) -> I256 {
+    x *= *ONE_18;
+
+    let z = ((x - *ONE_36) * *ONE_36) / (x + *ONE_36);
+    let z_squared = (z * z) / *ONE_36;
+
+    let mut num = z;
+    let mut series_sum = num;
+
+    for i in (3..=15).step_by(2) {
+        num = (num * z_squared) / *ONE_36;
+        series_sum += num / i.into();
+    }
+
+    series_sum * 2.into()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use regex::Regex;
+
+    // Compare computed constants with smart contract code.
```

Comment:
Could we also link to the corresponding smart contract tests here?

Reply:
At the time I wrote this, there were no public tests to read from. Here is a short test file from 5 days ago by Balancer, but I don't think it adds much: 
https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/solidity-utils/test/LogExpMath.test.ts